homeassistant/components/mqtt/config_flow.py

```diff
@@ -721,6 +853,294 @@ def _validate(
+        for field in [
+            form_field
+            for form_field in data_schema.schema
+            if form_field in component_data and form_field not in user_input
+        ]:
+            component_data.pop(field)
+        component_data.update(user_input)
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> SubentryFlowResult:
+        """Add a subentry."""
+        self._subentry_data = MqttSubentryData(device=MqttDeviceData(), components={})
+        return await self.async_step_device()
+
+    async def async_step_reconfigure(
+        self, user_input: dict[str, Any] | None = None
+    ) -> SubentryFlowResult:
+        """Reconfigure a subentry."""
+        reconfigure_subentry = self._get_reconfigure_subentry()
+        self._subentry_data = cast(
+            MqttSubentryData, deepcopy(dict(reconfigure_subentry.data))
+        )
+        return await self.async_step_summary_menu()
+
+    async def async_step_device(
+        self, user_input: dict[str, Any] | None = None
+    ) -> SubentryFlowResult:
+        """Add a new MQTT device."""
+        errors: dict[str, str] = {}
+        validate_field("configuration_url", cv.url, user_input, errors, "invalid_url")
+        if not errors and user_input is not None:
+            self._subentry_data[CONF_DEVICE] = cast(MqttDeviceData, user_input)
+            if self.source == SOURCE_RECONFIGURE:
+                self._dirty = True
+                return await self.async_step_summary_menu()
+            return await self.async_step_entity()
+
+        data_schema = self.add_suggested_values_to_schema(
+            MQTT_DEVICE_SCHEMA,
+            self._subentry_data[CONF_DEVICE] if user_input is None else user_input,
+        )
+        return self.async_show_form(
+            step_id=CONF_DEVICE,
+            data_schema=data_schema,
+            errors=errors,
+            last_step=False,
+        )
+
+    async def async_step_entity(
+        self, user_input: dict[str, Any] | None = None
+    ) -> SubentryFlowResult:
+        """Add or edit an mqtt entity."""
+        errors: dict[str, str] = {}
+        device_name = self._subentry_data[CONF_DEVICE][CONF_NAME]
+        data_schema_fields = COMMON_ENTITY_FIELDS
+        data_schema = data_schema_from_fields(data_schema_fields)
+        if user_input is not None:
+            validate_user_input(user_input, data_schema_fields, errors)
+            if not errors:
+                if self._component_id is None:
+                    self._component_id = uuid4().hex
+                self._subentry_data["components"].setdefault(self._component_id, {})
+                self.update_component_fields(data_schema, user_input)
+                return await self.async_step_mqtt_platform_config()
+            data_schema = self.add_suggested_values_to_schema(data_schema, user_input)
+        elif self.source == SOURCE_RECONFIGURE and self._component_id is not None:
```

Comment:
Is it possible that `self.source == SOURCE_RECONFIGURE` and `self._component_id` is None?
If not, this should be rewritten IMO:
```
elif self.source == SOURCE_RECONFIGURE:
    if TYPE_CHECKING:
        assert self._component_id is not None
    ...
```